### PR TITLE
Live cell changes

### PR DESCRIPTION
**Pull request by Dalantren**

### CODE REVIEW

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<ng-container [ngSwitch]="type">
+    <div class="live-cell {{ resultClass }} {{ [resultClass, type].join('-') }}" *ngSwitchCase="'currency'">
```
Comment (pavpka) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: maybe we should add template variable and set css classes in the component?

File: build.serve.js [hunk 2]
```diff
@@ -2,5 +2,5 @@
 
 const { spawn } = require('child_process');
 
-spawn('ng', ['serve', '--open'], { shell: true, stdio: 'inherit' });
+spawn('ng', ['serve', '--open' ], { shell: true, stdio: 'inherit' });
```
Comment (klumba12) [inline build.serve.js]: revert the change please

File: src/lib/main/core/body/body-core.component.html [hunk 3]
```diff
@@ -39,7 +39,8 @@
 			<td *ngFor="let $column of $view.body.render.columns($row, $table.pin, $rowIndex); index as $columnIndex; trackBy: columnId"
 				[attr.rowspan]="$view.body.render.rowspan($row, $column, $rowIndex, $columnIndex)"
 				[attr.colspan]="$view.body.render.colspan($row, $column, $rowIndex, $columnIndex)">
-				<ng-container [q-grid-core-td]="$column">
+				<ng-container 	[q-grid-core-td]="$column"
+								[liveValue]="$view.body.render.getValue($row, $column.model, $rowIndex, $columnIndex)">
```
Comment (klumba12) [inline src/lib/main/core/body/body-core.component.html]: `liveValue` -> `q-grid-core-value`

File: src/lib/main/core/body/td-core.directive.ts [hunk 4]
```diff
@@ -41,7 +45,18 @@ export class TdCoreDirective implements Td, OnInit, OnDestroy {
 		link(this.viewContainerRef, this);
 	}
 
-	mode(value: 'view' | 'edit') {
+	ngOnChanges(changes: SimpleChanges) {
+		for (const propName of Object.keys(changes)) {
```
Comment (klumba12) [inline src/lib/main/core/body/td-core.directive.ts]: just get necessary changes, don't go through the collection, it's performance critical

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	selector: 'q-grid-live-cell',
+	templateUrl: './live-cell.tpl.html',
+	providers: [TemplateHostService],
+	changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class LiveCellComponent implements OnInit {
+
+	@Input() cell: any;
+	@Input() duration = 500;
+	@Input() upClass = 'positive-change';
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: options bellow are under consideration,

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	@Input() downClass = 'negative-change';
+	@Input() showArrow = true;
+	@Input() showDifference = true;
+	@Input() showPrev = false;
+
+	difference = 0;
+	resultClass = '';
+	type = '';
+
+	constructor() {
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: get rid of empty expressions

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	ngOnInit() {
+		if (!this.cell.changes) {
+			this.cell.mode('view');
+			return;
+		}
+		this.type = this.cell.column.type;
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: don't persist this kind of props
just get them when required const { type } = this.cell.column;

File: src/lib/plugins/live-cell/live-cell.tpl.html (hunk 1 above)
Comment (klumba12) [reply src/lib/plugins/live-cell/live-cell.tpl.html]: use ngClass if possible

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        <span *ngIf="showArrow" class="live-cell__arrow">
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: all classes should start with q-grid

we don't use BEM get rid of __

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+        <span *ngIf="showDifference" class="live-cell__difference">{{ difference | qGridCurrency: cell.column.code}}</span>
+        <span *ngIf="showPrev" class="live-cell__previous">{{ cell.changes.liveValue.previousValue | qGridCurrency: cell.column.code}}</span>
+        <span class="live-cell__current">{{ cell.label | qGridCurrency: cell.column.code}}</span>
+    </div>
+
+    <div class="{{ resultClass }} {{ [resultClass, type].join('-') }}" *ngSwitchCase="'text'">
+        <span class="live-cell__current">{{ cell.label }}</span>
+        <span *ngIf="showPrev" class="live-cell__previous">{{ cell.changes.liveValue.previousValue }}</span>
+        <span *ngIf="showDifference" class="live-cell__difference">{{ difference }}</span>
+        <span *ngIf="showArrow" class="live-cell__arrow">
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: showArray showDifference can be controlled by user css don't need these variables in component and template

File: src/examples/live-data-basic/example-live-data-basic.component.html [hunk 10]
```diff
@@ -1,20 +1,31 @@
...
-		<q-grid-column key="metal" [width]="190"></q-grid-column>
+		<q-grid-column key="metal" [width]="290"></q-grid-column>
 		<q-grid-column key="ldn1" title="LDN" [width]="70"></q-grid-column>
-		<q-grid-column key="bid" [width]="100" type="currency"></q-grid-column>
-		<q-grid-column key="ask" [width]="100" type="currency"></q-grid-column>
+		<q-grid-column key="ask" [width]="250">
+			<ng-template for="change" let-$cell>
+				<q-grid-live-cell
+					[cell]="$cell"
+					duration="2000">
```
Comment (klumba12) [inline src/examples/live-data-basic/example-live-data-basic.component.html]: `[duration]="2000"`

File: src/examples/live-data-basic/example-live-data-basic.component.html [hunk 11]
```diff
@@ -1,20 +1,31 @@
...
+		</q-grid-column>
 		<q-grid-column key="ldn2" title="LDN" [width]="70"></q-grid-column>
-		<q-grid-column key="last" [width]="100" type="currency" [canSort]="false"></q-grid-column>
 		<q-grid-column key="ldn3" title="LDN" [width]="70"></q-grid-column>
-		<!-- <q-grid-column title="+/-" key="diff" type="number" [width]="70" [value]="diff" [canSort]="false" [canFilter]="false"></q-grid-column> -->
+		<q-grid-column key="last" type="currency" [width]="270">
+			<ng-template for="change" let-$cell>
+				<q-grid-live-cell
+					[cell]="$cell"
+					duration="3000">
```
Comment (klumba12) [inline src/examples/live-data-basic/example-live-data-basic.component.html]: [duration]

File: src/examples/live-data-basic/example-live-data-basic.component.html [hunk 12]
```diff
@@ -1,20 +1,31 @@
...
+				</q-grid-live-cell>
+			</ng-template>
+		</q-grid-column>
 		<q-grid-column key="high" [width]="100" type="currency"></q-grid-column>
 		<q-grid-column key="low" [width]="100" type="currency"></q-grid-column>
 		<q-grid-column key="volume" [width]="100" type="currency"></q-grid-column>
 		<q-grid-column key="settle" [width]="100" type="currency"></q-grid-column>
 		<q-grid-column key="ldn4" title="LDN" [width]="70"></q-grid-column>
-		<q-grid-column key="previous" [isVisible]="false"></q-grid-column>			
+		<q-grid-column key="previous" [isVisible]="false"></q-grid-column>
```
Comment (klumba12) [inline src/examples/live-data-basic/example-live-data-basic.component.html]: remove this column

File: src/examples/live-data-basic/example-live-data-basic.component.scss [hunk 13]
```diff
@@ -1,8 +1,29 @@
-.q-grid-last.negative {
-    color: #E46859 !important;
- }
- 
- .q-grid-last.positive {
-    color: #228B0F !important;
- }
- 
\ No newline at end of file
+span {
```
Comment (klumba12) [inline src/examples/live-data-basic/example-live-data-basic.component.scss]: what is span?

File: src/examples/live-data-basic/example-live-data-basic.component.scss [hunk 14]
```diff
@@ -1,8 +1,29 @@
...
+   display: inline-block;
+   padding: 5px;
+}
+
+.q-grid-live-cell {
```
Comment (klumba12) [inline src/examples/live-data-basic/example-live-data-basic.component.scss]: all this styles should go to the q-grid css by deafult,
not in example

File: src/lib/main/core/body/td-core.directive.ts [hunk 15]
```diff
@@ -15,11 +16,14 @@ const classify = TdCtrl.classify;
...
 })
-export class TdCoreDirective implements Td, OnInit, OnDestroy {
+export class TdCoreDirective implements Td, OnInit, OnDestroy, OnChanges {
 	private $implicit = this;
+	@Input('q-grid-core-value') private actualValue: any;
 
 	@Input('q-grid-core-td') columnView: ColumnView;
+
 	element: HTMLElement = null;
+	changes: SimpleChange = null;
```
Comment (klumba12) [inline src/lib/main/core/body/td-core.directive.ts]: Cool!

File: src/lib/main/core/body/td-core.directive.ts [hunk 16]
```diff
@@ -41,7 +45,17 @@ export class TdCoreDirective implements Td, OnInit, OnDestroy {
 		link(this.viewContainerRef, this);
 	}
 
-	mode(value: 'view' | 'edit') {
+	ngOnChanges(changes: SimpleChanges) {
+
```
Comment (klumba12) [inline src/lib/main/core/body/td-core.directive.ts]: do not make extra new lines between method declaration and body

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+import { TdCoreDirective } from 'lib/main/core/body/td-core.directive';
+
+@Component({
+	selector: 'q-grid-live-cell',
+	templateUrl: './live-cell.tpl.html',
+	providers: [TemplateHostService],
+	changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class LiveCellComponent implements OnInit {
+
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: remove New Line

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+	constructor() {
+	}
+
+	ngOnInit() {
+		if (!this.cell.changes) {
+			this.cell.mode('view');
+			return;
+		}
+		this.difference = this.diff(this.cell.changes);
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: I think we need to propagate to live template `changes` not `difference`, for instance for text template difference doesn't make any sense.

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+	@Input() cell: TdCoreDirective;
+	@Input() duration = 500;
+
+	difference = 0;
+
+	constructor() {
+	}
+
+	ngOnInit() {
+		if (!this.cell.changes) {
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: I believe we can not enter here if there is no changes, get rid of this check please.
Other way is to throw exception.

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+			this.cell.mode('view');
+			return;
+		}
+		this.difference = this.diff(this.cell.changes);
+
+		setTimeout(() => {
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: Use it outside angular zone. 
zone.runOutsideAngular

File: src/lib/plugins/live-cell/live-cell.module.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
+import { NgModule, } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { TemplateModule } from '../../template/template.module';
+import { LiveCellComponent } from './live-cell.component';
+import { PipeModule } from '../../../lib/pipes/pipe.module';
+
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.module.ts]: remove new line

File: src/lib/plugins/live-cell/live-cell.component.ts (hunk 18 above)
Comment (klumba12) [reply src/lib/plugins/live-cell/live-cell.component.ts]: Also getDifference(changes) method can be added that calculates the value, not persist.

File: src/lib/assets/live-cell.scss [hunk 22]
```diff
@@ -0,0 +1,39 @@
+.q-grid-live-cell {
```
Comment (klumba12) [inline src/lib/assets/live-cell.scss]: rename file to `live.cell.scss`

File: src/lib/main/core/body/td-core.directive.ts [hunk 23]
```diff
@@ -61,16 +75,23 @@ export class TdCoreDirective implements Td, OnInit, OnDestroy {
 				this.cd.detectChanges();
 				break;
 			}
+			case 'change': {
+				this.element.classList.add(`${GRID_PREFIX}-change`);
```
Comment (klumba12) [inline src/lib/main/core/body/td-core.directive.ts]: need to remove this class on mode `view`

File: src/lib/main/core/body/td-core.directive.ts [hunk 24]
```diff
@@ -61,16 +75,23 @@ export class TdCoreDirective implements Td, OnInit, OnDestroy {
...
+			}
 			default:
 				throw new AppError('td.core', `Invalid mode ${value}`);
 		}
 	}
 
 	get value() {
-		const { column, row, rowIndex, columnIndex } = this;
-		return this.$view.body.render.getValue(row, column, rowIndex, columnIndex);
+		return this.actualValue;
```
Comment (klumba12) [inline src/lib/main/core/body/td-core.directive.ts]: It looks like no need to persist value in the own field 
just use
`this.changes && this.changes.currentValue`

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 25]
```diff
@@ -0,0 +1,44 @@
+import { Component, Input, OnInit, SimpleChange, ChangeDetectionStrategy, NgZone } from '@angular/core';
+import { TemplateHostService } from '../../template/template-host.service';
+import { TdCoreDirective } from 'lib/main/core/body/td-core.directive';
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: please fix path, it should be relative. `../..`

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+	@Input() duration = 500;
+
+	changes: any;
+
+	constructor(private zone: NgZone) {
+	}
+
+	ngOnInit() {
+		if (!this.cell.changes) {
+			throw new AppError('q-grid-live-cell', 'Changes is not defined in live-cell.component.ts');
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: throw new AppError('live-cell.component', 'No changes found');

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+		}
+
+		this.changes = this.getDifference(this.cell.changes);
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: It should not be calculated here, just call it from template

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+		this.zone.runOutsideAngular(() => {
+			setTimeout(() => {
+				this.cell.mode('view');
+			}, this.duration);
+		});
+	}
+
+	getDifference(value: SimpleChange) {
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: we need to use different method for numbers strings date etc.

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 29]
```diff
@@ -0,0 +1,31 @@
+<ng-container [ngSwitch]="cell.column.type">
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: this html should go under theme and for each column type we need to have own template like for body cell templates.

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+    <div *ngSwitchCase="'currency'"
+    [ngClass]="{ 'q-grid-live-cell' : true,
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: use @HostBinding for q-grid-live-cell and q-grid-live-cell-currency.

File: src/lib/plugins/live-cell/live-cell.component.ts (hunk 28 above)
Comment (Dalantren) [reply src/lib/plugins/live-cell/live-cell.component.ts]: yep, so that's why i use switch-case in this method, depends on type of variable. Is it wrong?
we can get difference, when we working with numbers. In case of strings (dates, and so on) we just get previous value. Or how should i determine the type of input data? For example, for dates it's just a number. And how we should calculate difference for objects (arrays), Boolean, etc.

File: src/lib/plugins/live-cell/live-cell.component.ts (hunk 27 above)
Comment (Dalantren) [reply src/lib/plugins/live-cell/live-cell.component.ts]: but i use this variable a lot of time in the template. Why should i calculate the value of a function many times, if I can calculate it once and put it in a variable?

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 31]
```diff
@@ -0,0 +1,65 @@
+import { Component, Input, OnInit, SimpleChange, ChangeDetectionStrategy,
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: Please format it like that 
import {
 Component,
 Input,
 ...
}

File: src/lib/plugins/live-cell/live-cell.component.ts [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+	templateUrl: './live-cell.tpl.html',
+	providers: [TemplateHostService],
+	changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class LiveCellComponent implements OnInit {
+	@Input() cell: TdCoreDirective;
+	@Input() duration = 500;
+
+	changes: any;
+	type: string;
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.component.ts]: Please remove this.

File: src/lib/plugins/live-cell/live-cell.module.ts [hunk 33]
```diff
@@ -0,0 +1,21 @@
...
+@NgModule({
+	declarations: [
+		LiveCellComponent,
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.module.ts]: Fix formatting in this file

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 34]
```diff
@@ -0,0 +1,45 @@
+<ng-container *ngTemplateOutlet="getTemplate()"></ng-container>
+<!--Templates: -->
+<ng-template #currency>
+    <div [ngClass]="{'q-grid-live-cell-container': true,
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: {
  'q-grid....

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 35]
```diff
@@ -0,0 +1,45 @@
...
+<ng-template #currency>
+    <div [ngClass]="{'q-grid-live-cell-container': true, 
+                    'q-grid-live-cell-up': changes > 0,
+                    'q-grid-live-cell-down': changes < 0
+            }">
+        <span [ngClass]="'q-grid-live-cell-arrow'">
+            <i *ngIf="changes > 0" class="q-grid-icon q-grid-asc">&uarr;</i>
+            <i *ngIf="changes < 0" class="q-grid-icon q-grid-desc">&darr;</i>
+        </span>
+        <span [ngClass]="'q-grid-live-cell-diff'">{{ changes | qGridCurrency: cell.column.code}}</span>
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: just class="q-grid-live-cell"

File: src/lib/plugins/live-cell/live-cell.component.ts (hunk 28 above)
Comment (klumba12) [reply src/lib/plugins/live-cell/live-cell.component.ts]: Let's not persist it, and propagate to the template `getDiff()` method, we can apply cache in future if face with performance issues.

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 36]
```diff
@@ -0,0 +1,45 @@
+<ng-container *ngTemplateOutlet="getTemplate()"></ng-container>
+<!--Templates: -->
+<ng-template #currency>
+    <div [ngClass]="{'q-grid-live-cell-container': true, 
+                    'q-grid-live-cell-up': changes > 0,
+                    'q-grid-live-cell-down': changes < 0
+            }">
+        <span [ngClass]="'q-grid-live-cell-arrow'">
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: not sure why we need here this span, please remove it

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 37]
```diff
@@ -0,0 +1,45 @@
...
+        <span [ngClass]="'q-grid-live-cell-curr'">{{ cell.label | qGridCurrency: cell.column.code}}</span>
+    </div>
+</ng-template>
+
+<ng-template #number>
+        <div [ngClass]="{'q-grid-live-cell-container': true,
+                        'q-grid-live-cell-up': changes > 0,
+                        'q-grid-live-cell-down': changes < 0
+                }">
+        <span [ngClass]="'q-grid-live-cell-curr'">{{ cell.label }}</span>
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: if we use label here, in cell actualValue should be replaced by actualLabel

File: src/lib/plugins/live-cell/live-cell.tpl.html [hunk 38]
```diff
@@ -0,0 +1,45 @@
...
+        <span [ngClass]="'q-grid-live-cell-diff'">{{ changes }}</span>
+        <span [ngClass]="'q-grid-live-cell-arrow'">
+            <i *ngIf="changes > 0" class="q-grid-icon q-grid-asc">&uarr;</i>
+            <i *ngIf="changes < 0" class="q-grid-icon q-grid-desc">&darr;</i>
+        </span>
+    </div>
+</ng-template>
+
+<ng-template #time>
+    <div [ngClass]="'q-grid-live-cell-container'">
```
Comment (klumba12) [inline src/lib/plugins/live-cell/live-cell.tpl.html]: don't think we need this div here, it doesn't give us any semantic or function
